paddle/fluid/memory/detail/system_allocator.cc

```diff
@@ -79,7 +79,18 @@ void* GPUAllocator::Alloc(size_t& index, size_t size) {
   // if size is 0.  We just make sure it does.
   if (size <= 0) return nullptr;
   void* p;
+  int prev_id;
+  cudaGetDevice(&prev_id);
+  if (prev_id != gpu_id_) {
+    cudaSetDevice(gpu_id_);
+  }
+
   cudaError_t result = cudaMalloc(&p, size);
+
+  if (prev_id != gpu_id_) {
+    cudaSetDevice(prev_id);
+  }
+
```

Comment:
During the runtime, operators will be calling `cudaSetDevice`, what is the purpose of these `cudaSetDevice`?

Reply:
I think the Allocator should be self-contained. It should make sure the device is correct and do not change the global devices.